src/main/java/uk/gov/ch/developer/docs/controller/developer/UserCallbackController.java

```diff
@@ -1,76 +1,48 @@
-import org.apache.commons.codec.binary.Base64;
+import com.nimbusds.jose.Payload;
+import net.minidev.json.JSONObject;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Controller;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestParam;
-import com.nimbusds.jose.JOSEException;
-import com.nimbusds.jose.JWEObject;
-import com.nimbusds.jose.Payload;
-import com.nimbusds.jose.crypto.DirectDecrypter;
-import net.minidev.json.JSONObject;
-import net.minidev.json.JSONValue;
-import uk.gov.ch.developer.docs.session.SessionService;
-import uk.gov.companieshouse.environment.EnvironmentReader;
+import org.springframework.web.bind.annotation.ResponseBody;
+import uk.gov.ch.oauth.IIdentityProvider;
+import uk.gov.ch.oauth.IOauth;
 import uk.gov.companieshouse.logging.Logger;
 import uk.gov.companieshouse.logging.LoggerFactory;
-import uk.gov.companieshouse.session.Session;
-import uk.gov.companieshouse.session.SessionKeys;
 
 @Controller
 @RequestMapping("${callback.url}")
 public class UserCallbackController {
 
     private static final Logger LOGGER = LoggerFactory.getLogger("docs.developer.ch.gov.uk");
-    private final String base64Key;
-    private final String redirectUri;
-
     @Autowired
-    private SessionService sessionService;
-
-    public UserCallbackController(EnvironmentReader reader) {
-        this.base64Key = reader.getMandatoryString("OAUTH2_REQUEST_KEY");
-        this.redirectUri = reader.getMandatoryString("REDIRECT_URI");
-    }
+    IIdentityProvider identityProvider;
+    @Autowired
+    private IOauth oauth;
 
     @GetMapping
-    public String getParams(HttpServletRequest httpServletRequest,
-            HttpServletResponse httpServletResponse, @RequestParam("state") final String state,
-            @RequestParam("code") final String code)
-            throws ParseException, JOSEException, net.minidev.json.parser.ParseException {
-
-        Session sessionCallback = sessionService.getSessionFromContext();
-
-        Map<String, Object> sessionMap = sessionCallback.getData();
-        String nonceSession = sessionMap.get(SessionKeys.NONCE.getKey()).toString();
-
-        String nonceState = decryptState(state);
-
-        if (nonceState.equals(nonceSession)) {
-            LOGGER.debug("Nonce values are equal");
+    @ResponseBody
+    public String getCallback(@RequestParam("state") String state,
+            @RequestParam("code") String code) {
+        LOGGER.trace("Code:" + code);
+        LOGGER.trace("State:" + state);
+
+        final String returnedNonce = getNonceFromState(state);
+        if (!oauth.oauth2VerifyNonce(returnedNonce)) {
+            LOGGER.error("Invalid nonce value in state during oauth2 callback");
+            return "redirect:/";
+        } else {
+            return ("redirect:" + identityProvider.getRedirectUri());
```

Comment:
This is incorrect, we don't want to redirect back to the callback once nonce values have been asserted to be equal. In face we could probably do away with the else statement as this redirect should be the last redirect after all the logic of callback has been implemented. At the moment I have redirected it back to home page, there is a ticket in the backlog to look at proper redirection back to the page where the user has initiated the sign-in request from.

Reply:
@Zain-Abbas1 Would you mind correcting the redirect in the next iteration of this code?

Reply:
Would be good to put a TODO there even if you don't change it yourself. I know that I'd forget if I picked this up.

Reply:
Yes

Reply:
I have added a new redirectUriPage which is just the redirect back to the home page for now.